app/engine/ai.py

```diff
@@ -37,125 +34,304 @@ def __init__(
         self.is_ready = True
         self.logged_in = True
         self.is_bot = True
+        self.cards_queue = []
+        self.logger = logging.getLogger('AI')
+
+    def debug_log(self, message: str) -> None:
+        self.logger.info(f'{self.element.capitalize()} {message}')
 
     @delay(0.25, 1.5)
     def confirm_move(self) -> None:
         if self.is_ready:
             return
 
-        confirm = GameObject(
-            self.game,
-            'ui_confirm',
-            x_offset=0.5,
-            y_offset=1.05
-        )
-
-        confirm.x = self.ninja.x
-        confirm.y = self.ninja.y
+        confirm = GameObject(self.game, 'ui_confirm', x_offset=0.5, y_offset=1.05)
+        confirm.x, confirm.y = self.ninja.x, self.ninja.y
         confirm.place_object()
         confirm.place_sprite(confirm.name)
         confirm.play_sound('SFX_MG_2013_CJSnow_UIPlayerReady_VBR8')
         self.is_ready = True
 
     @delay(0.5, 3)
     def select_move(self) -> None:
-        # Check for k.o. state
         if self.ninja.hp <= 0:
-            if not self.member_card:
-                self.confirm_move()
-                return
-
-            self.member_card.place()
+            self.handle_knockout()
             self.confirm_move()
             return
 
-        # Check for k.o. allies
         for ninja in self.game.ninjas:
-            if not ninja.hp <= 0:
-                continue
+            if self.should_revive_ally(ninja):
+                self.debug_log(f"reviving ally at position: ({ninja.x}, {ninja.y})")
+                self.select_target(ninja.x, ninja.y)
+                break
+        else:
+            self.action_strategy()
 
-            if ninja.client.member_card:
-                continue
+        self.confirm_move()
 
-            if self.is_ninja_getting_revived(ninja):
-                continue
+    def handle_knockout(self) -> None:
+        if self.member_card:
+            self.debug_log("is using a member card for revival.")
+            self.member_card.place()
 
-            if not self.can_heal_ninja(ninja):
-                continue
+    def is_ninja_selected(self, ninja: Ninja) -> bool:
+        return any(ninja.selected_object == ninja for ninja in self.game.ninjas)
```

Comment:
Please check if this section works correctly. The `ninja` parameter might get overridden inside the for-loop.